src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -131,6 +141,55 @@ internal void Write(byte[] buffer, int offset, int count)
             ProcessWrite(buffer, offset, count, null);
         }
 
+        internal Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (!_sslState.InnerStream.CanWrite) return Task.FromException(new NotSupportedException(SR.NotSupported_UnwritableStream));
+            // If cancellation was requested, bail early with an already completed task.
+            // Otherwise, return a task that represents the Begin/End methods.
+            return cancellationToken.IsCancellationRequested
+                        ? Task.FromCanceled(cancellationToken)
+                        : WriteAsyncImpl(buffer, offset, count, cancellationToken);
+        }
+
+        private async Task WriteAsyncImpl(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            var semaphore = EnsureAsyncActiveWriteSemaphoreInitialized();
+            await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+            InitiateWrite(buffer, offset, count, null);
+
+            bool failed = false;
+            try
+            {
+                await StartWritingAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+
+                if (Interlocked.Exchange(ref _nestedWrite, 0) == 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "WriteAsyncImpl"));
```

Comment:
Is this correct?  invalidendcall isn't relevant for WriteAsync, and WriteAsyncImpl isn't publically exposed, so that name doesn't mean anything to a caller.  Is this actually dead code, or if not, in what situations is this possible?